src/dialog/llm/abstract_llm.py

```diff
@@ -79,8 +80,11 @@ def process(self, input: str):
         """
         processed_input = self.preprocess(input)
         self.generate_prompt(processed_input)
-        output = self.llm({
-            "user_message": processed_input,
-        })
+        if self.fallback:
```

Comment:
This behavior doesn't make sense. We are checking to see if the variable fallback exists not if the context is really a context that the LLM doesn't have the knowledge about.

Reply:
@vmesel, I believe @lgabs's point concerns returning a fallback without submitting it as input to the LLM. In this scenario, the LLM model won't be invoked if no document is retrieved. It makes sense to check the self.fallback because the class is instantiated with every message from the customer. The last suggestion seems more elegant, where we adjust the final prompt within the process method's concluding call.

Reply:
@vmesel this implementation was made on abstract llm, i.e., it is design to work on default llm and other abstract llm subclasses. We've checked that already, but you can verify yourself asking a out of scope question to check that locally, and than ask some question inside the knowledge base scope to check this as well.

This is how it works in this branch (see diffs for better comparison):

- every time some endpoint instantiate the default llm classe via `get_llm_class`, fallback attribute is set to `None`, just like others like `prompt`. This happens on [abstract_llm.py](https://github.com/lgabs/dialog/blob/luan/fix-fallback-behavior/src/dialog/llm/abstract_llm.py#L22).
- in the current design, AbstractLLM calls `generate_prompt` [here](https://github.com/lgabs/dialog/blob/luan/fix-fallback-behavior/src/dialog/llm/abstract_llm.py#L82), which in default llm is responsible for retrieval, and so I've put an if-else in case of no retrieved documents, setting the llm class attribute `fallback` to the default template, [here](https://github.com/lgabs/dialog/blob/luan/fix-fallback-behavior/src/dialog/llm/default.py#L48)
- finally, the "if-else" you've highlighted occurs after the prompt generation and checks for the `self.fallback`: it will be `None` when some docs were retrieved inside prompt generation, otherwise it will be the fallback template.